### ISSUE TITLE
If pattern match is not correct when cast  without if ` bool casted = s is string s1;`

### ISSUE THREAD

**Issue by MauNguyenVan** [dotnet/roslyn#71955]
**Brief description:**
My code:
`object? o = null;`
` bool casted = o is string s1;` it is legal code, then, when use `s1` VS will always emits an error when use `s1`. Although VS exposes and suggests all string API member   

### IMO: 
```cs 
object? o = null;
bool casted = o is string s1;
if (casted)
{
}
```
it should be  same with 
```cs 
if(o is string s1)
{
}
```
Also same with `is not` pattern    

**Languages applicable:**
C# 12

**Code example that the analyzer should report:**
[code in sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxAvsi0o0QJYB2ALgKYAnbgEMANphjoAKvwDOvAJJyAygFsRg3thEKODJMSOYKshXX1GCh47YD2AIwBW/AMa8A/OjvoAvOm4AVzExAG5LW28/dAAiHQV0AAcRXgFhdA1XAAseAHN0AHdBO258rKF+GPCbSOIHOzsJV10BYGifTjlMOAAGdDk4atqjTgAzdGpmhX5gWgjba2HIgHplqDgATkmWmYA6aTsVXkE8ulohpeN1rYHdgBl+Ut4s8/njNkv0fjE5fjerf6Ra7UW4HACqiUSQjOFyWH0uwIGrxq730TCAA===)
```cs
    internal class TestIsSmartCast
    {
        void Test()
        {
            object? o = null;
            o = "Cast pattern maching wrong here";
            bool casted = o is string s1; // legal
            if (casted)
            {
                //Console.WriteLine(casted.ToString());
                Console.WriteLine(s1.Length);
            }
            else
            {
                Console.WriteLine(s1.ToUpper());
            }
            Console.WriteLine(s1);
        }
    }
```
![image](https://github.com/dotnet/roslyn/assets/48544354/ebe0c9e4-f6ca-4f25-9de8-b1f06758dcc4)

**Comment by DoctorKrolic** [comment]
That's currently by design since compiler doesn't track relations between variables. In this specific case it means that there is no way to tell, that if `casted` is `true` then `s1` is initialized

**Comment by CyrusNajmabadi** [comment]
By design with how the language currently works. If you would like a language change, that should start with a discussion at csharplang

**Comment by CyrusNajmabadi** [comment]
> It will be better if language support this syntax

Totally understand.  But please start a discussion on dotnet/csharplang if you'd like to pursue that :)

**Comment by huoyaoyuan** [comment]
> but canot use s1 to do any thing.

The variable is only considered assigned in the range that the pattern matching is considered succussed. That's just the `if` directly contains the match statement.

> Although you can see explicite type when hover on it, get member when type with suggestion from VS

It's considered declared but not assigned in other range, similar to a solely variable declaration `string s1;`. If you assign to it later, the it is assigned.

**Comment by huoyaoyuan** [comment]
> because obj is allways assigned

The compiler (and language) only do such propagation if obj is `const`:

```csharp
public void M1()
{
    const string s = "abc";
    bool x = s is string s1;
    Console.WriteLine(s1); // no error
}

public void M2()
{
    string s = "abc";
    bool x = s is string s1;
    Console.WriteLine(s1); // error
}
```

The assignment of variables are not tracked, because it needs more complexity to determine the effective assignment.
